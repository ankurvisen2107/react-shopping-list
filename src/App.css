.App {
  text-align: center;
  margin-top: 1%;
}
.item-list {
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(300px, 1fr));
  gap: 20px;
}
