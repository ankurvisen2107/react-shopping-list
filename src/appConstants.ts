import {Category} from "./appModels";

export const BASE_URL = 'http://localhost:8000';
export const ITEMS_URL = `${BASE_URL}/items`;
export const CATEGORIES_URL = `${BASE_URL}/categories`;

export const ALL_CATEGORIES_ID = 0;

export const NO_DATA_FOUND = 'No Data found..';

export const LOADING_ITEMS = 'Loading Items...'

export const ALL_CATEGORIES: Category = {
    id: ALL_CATEGORIES_ID,
    name: 'all (すべて)'
}