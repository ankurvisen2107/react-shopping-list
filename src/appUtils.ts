import {Category, Item} from "./appModels";
import {ALL_CATEGORIES} from "./appConstants";

export function filterItems(items : Item[], query: string, categoryId: number) {
    const filterItems = items.filter(item =>
        item.name.toLowerCase().startsWith(query.toLowerCase())
    );
    if (categoryId>0){
        return filterItems.filter(filterItems => filterItems.category_id == categoryId);
    }
    return filterItems;
}

export function getAllCategories(categories: Category[]): Category[] {
        const allCategories = [...categories];
        allCategories.unshift(ALL_CATEGORIES);
        return allCategories;
}
