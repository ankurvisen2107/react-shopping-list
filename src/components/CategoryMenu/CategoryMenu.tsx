import React, {EventHandler, FC, useEffect, useState} from 'react';
import {CATEGORIES_URL} from "../../appConstants";
import axios from "axios";
import {Category} from "../../appModels";
import {getAllCategories} from "../../appUtils";

interface Props {
    activeTab: number,
    // @ts-ignore
    onMenuClick: EventHandler<number>
}

const CategoryMenu: FC<Props> = ({onMenuClick, activeTab}) => {
    const [categories, setCategories] = useState<Category[]>([]);
    useEffect(() => {
        axios.get(CATEGORIES_URL)
            .then(response => {
                // @ts-ignore
                setCategories(Object.values(response.data)[0]);
            })
            .catch(error => {
                console.error('Error fetching data:', error);
            });
    }, []);
    const finalCategories = getAllCategories(categories);
    return (
        <div className="tab-component">
            <div className="categories">
                {
                    !!finalCategories && finalCategories.map((category: Category) => {
                        return (
                            <div key={category.id}
                                 className={`tab ${activeTab === category.id ? 'active' : ''}`}
                                 onClick={() => onMenuClick(category.id)}
                            >{category.name}
                            </div>)
                    })
                }
            </div>
        </div>
    );
};

export default React.memo(CategoryMenu);
