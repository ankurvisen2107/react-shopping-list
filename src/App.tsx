import React, {Suspense, useCallback, useEffect, useState} from 'react';
import './appStyles.css'
import SearchBar from "./components/SearchBar/SearchBar";
import axios from "axios";
import {ITEMS_URL, LOADING_ITEMS} from "./appConstants";
import {filterItems} from "./appUtils";
import CategoryMenu from "./components/CategoryMenu/CategoryMenu";
import {Item} from "./appModels";
import {Route, Routes} from "react-router-dom";
import ItemDetail from "./components/ItemDetail/ItemDetail";
import {useMemo} from 'react';
import NotificationPanel from "./components/Generic/NotificationPanel";
const ItemListView = React.lazy(() => import('./components/ItemView/ItemListView'));

const App = () => {
    const [query, setQuery] = useState('');
    const [categoryId, setCategoryId] = useState(0);
    const [items, setItems] = useState<Item[]>([]);

    const handleChange = useCallback(
        (e: React.ChangeEvent<HTMLInputElement>) => {
            setQuery(e.target.value);
        },
        []
    );

    const handleClick = useCallback(
        (id: number) => {
            setCategoryId(id);
        },
        []
    );

    useEffect(() => {
        axios.get(ITEMS_URL)
            .then(response => {
                // @ts-ignore
                setItems(Object.values(response.data)[0]);
            })
            .catch(error => {
                console.error('Error fetching data:', error);
            });
    }, []);
    const filteredItems = useMemo(
        () => filterItems(items, query, categoryId),
        [items, query, categoryId]
    );
    return (
        <div className="App bg-gray">
            <Routes>
                <Route path="/"
                       element={<>
                           <SearchBar query={query} onChange={handleChange}/>
                           <CategoryMenu activeTab={categoryId} onMenuClick={handleClick}/>
                           <Suspense fallback={<NotificationPanel text={LOADING_ITEMS} />}>
                           <ItemListView itemList={filteredItems}/>
                           </Suspense>
                       </>}/>
                <Route path="/items/:id" element={<ItemDetail/>}/>
            </Routes>
        </div>
    );
}

export default App;
