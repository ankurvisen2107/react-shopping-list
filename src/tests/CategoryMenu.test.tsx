import {render, screen,} from "@testing-library/react";
import {BrowserRouter} from "react-router-dom";
import App from "../App";
import {filterItems, getAllCategories} from "../appUtils";
import {mockCategories, mockCategoryItem, mockItemList} from "./testConstants";
import CategoryMenu from "../components/CategoryMenu/CategoryMenu";
import {ALL_CATEGORIES} from "../appConstants";

test('Renders the landing page', () => {
    const func = jest.fn();
    render(<BrowserRouter><CategoryMenu activeTab={0} onMenuClick={func} /></BrowserRouter>);
});

describe('CategoryMenu component', () => {

    it('Renders All Category', () => {
        render(<BrowserRouter><App /></BrowserRouter>);
        const tab = screen.getByText(/all/);
        expect(tab).toBeInTheDocument();
    });

    it('should prepend All Category object to Category list', () => {
        const finalCategories = getAllCategories(mockCategories);
        expect(finalCategories[0]).toMatchObject(ALL_CATEGORIES);
    });

    it('should filter items correctly on basis of category id', function () {
        const mockFilteredValues = filterItems(mockItemList, mockCategoryItem.name, mockCategoryItem.category_id);
        expect(mockFilteredValues).toMatchObject([mockCategoryItem]);
    });

    it('matches snapshot', () => {
        const func = jest.fn();
        const { container } = render(<BrowserRouter><CategoryMenu activeTab={0} onMenuClick={func} /></BrowserRouter>);
        expect(container).toMatchSnapshot();
    });
});
