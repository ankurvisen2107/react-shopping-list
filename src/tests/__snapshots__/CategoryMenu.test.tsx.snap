// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryMenu component matches snapshot 1`] = `
<div>
  <div
    class="tab-component"
  >
    <div
      class="categories"
    >
      <div
        class="tab active"
      >
        all (すべて)
      </div>
    </div>
  </div>
</div>
`;
